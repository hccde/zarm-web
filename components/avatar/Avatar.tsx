import React, { Component, ReactEventHandler, HTMLAttributes, CSSProperties } from 'react';
import classnames from 'classnames';
import { PropsType, StateType } from './PropsType';
import Icon from '../icon';

class Avatar extends Component<PropsType & HTMLAttributes<HTMLSpanElement>, StateType> {
  static defaultProps: PropsType = {
    prefixCls: 'za-avatar',
    icon: '',
    shape: 'circle',
    src: '',
    alt: '',
  };

  state: StateType = {
    loadError: false,
  };

  onError: ReactEventHandler<HTMLImageElement> = (e) => {
    this.setState({ loadError: true });
    if (this.props.onError) {
      this.props.onError(e);
    }
  };

  render() {
    const {
      prefixCls,
      style,
      size,
      shape,
      icon,
      src,
      alt,
      children,
      className,
      ...others
    } = this.props;
    const { loadError } = this.state;

    const hasIcon = (icon && icon.trim() !== '');
    const hasImage = (src && src.trim() !== '');
    const hasString = (typeof children === 'string');
    const strLength = (typeof children === 'string' && children.length) || 0;

    const cls = classnames({
      [prefixCls!]: true,
      [className!]: className,
      [`${prefixCls}--${size}`]: typeof size === 'string' && size,
      [`${prefixCls}--${shape}`]: true,
      [`${prefixCls}--icon`]: hasIcon,
      [`${prefixCls}--image`]: hasImage && !loadError,
      [`${prefixCls}--string`]: hasString || (hasImage && loadError),
    });
    const inlineStyle: CSSProperties = (typeof size === 'number') ? {
      ...style,
      width: size,
      height: size,
      lineHeight: `${size}px`,
      fontSize: `${size / 2}px`,
    } : { ...style };

    const spanStyle: CSSProperties = {
      position: 'absolute',
      transform: `scale(${1 - 0.1 * (strLength - 1)})`,
      left: `calc(50% - ${4.5 * strLength}px )`,
    };

    return (
      <span
        style={inlineStyle}
        className={cls}
        {...others}
      >
        {hasImage && !loadError && <img src={src} alt={alt} onError={this.onError} />}
        {!hasImage && hasIcon && <Icon type={icon} />}
        {(!hasImage || (hasImage && loadError)) && !hasIcon && hasString && <span style={spanStyle}>{children}</span>}
      </span>
    );
  }
}

export default Avatar;
