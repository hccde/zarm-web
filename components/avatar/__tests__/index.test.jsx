import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Avatar from '../index';

describe('Avatar', () => {
  it('renders Avatars of different size correctly', () => {
    const wrapper = render(
      <div>
        <Avatar size={80} icon="user" />
        <Avatar size="xl" icon="user" />
        <Avatar size="lg" icon="user" />
        <Avatar icon="user" />
        <Avatar size="sm" icon="user" />
        <Avatar size="xs" icon="user" />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Avatars of different shape correctly', () => {
    const wrapper = shallow(
      <div>
        <Avatar size="circle" />
        <Avatar size="square" />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Avatars of different type correctly', () => {
    const wrapper = render(
      <div>
        <Avatar icon="user" />
        <Avatar>U</Avatar>
        <Avatar>USER</Avatar>
        <Avatar src="http://img95.699pic.com/element/40044/5588.png_860.png" alt="my avatar" onError={() => console.log('load error')} />
        <Avatar style={{ color: '#2db7f5', backgroundColor: 'rgb(249, 232, 8)' }}>U</Avatar>
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Avatars correctly when set src, icon and string', () => {
    const wrapper = render(
      <div>
        <Avatar src="http://img95.699pic.com/element/40044/5588.png_860.png" icon="user" >USER</Avatar>
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
