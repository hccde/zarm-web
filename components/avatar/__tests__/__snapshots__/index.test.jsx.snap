// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders Avatars correctly when set src, icon and string 1`] = `
<div>
  <span
    class="za-avatar za-avatar--circle za-avatar--icon za-avatar--image za-avatar--string"
  >
    <img
      alt=""
      src="http://img95.699pic.com/element/40044/5588.png_860.png"
    />
  </span>
</div>
`;

exports[`Avatar renders Avatars of different shape correctly 1`] = `
<div>
  <Avatar
    alt=""
    icon=""
    prefixCls="za-avatar"
    shape="circle"
    size="circle"
    src=""
  />
  <Avatar
    alt=""
    icon=""
    prefixCls="za-avatar"
    shape="circle"
    size="square"
    src=""
  />
</div>
`;

exports[`Avatar renders Avatars of different size correctly 1`] = `
<div>
  <span
    class="za-avatar za-avatar--circle za-avatar--icon"
    style="width:80px;height:80px;line-height:80px;font-size:40px"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--xl za-avatar--circle za-avatar--icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--lg za-avatar--circle za-avatar--icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--circle za-avatar--icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--sm za-avatar--circle za-avatar--icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--xs za-avatar--circle za-avatar--icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
</div>
`;

exports[`Avatar renders Avatars of different type correctly 1`] = `
<div>
  <span
    class="za-avatar za-avatar--circle za-avatar--icon"
  >
    <i
      class="za-icon za-icon-user theme-default"
    />
  </span>
  <span
    class="za-avatar za-avatar--circle za-avatar--string"
  >
    <span
      style="position:absolute;transform:scale(1);left:calc(50% - 4.5px )"
    >
      U
    </span>
  </span>
  <span
    class="za-avatar za-avatar--circle za-avatar--string"
  >
    <span
      style="position:absolute;transform:scale(0.7);left:calc(50% - 18px )"
    >
      USER
    </span>
  </span>
  <span
    class="za-avatar za-avatar--circle za-avatar--image"
  >
    <img
      alt="my avatar"
      src="http://img95.699pic.com/element/40044/5588.png_860.png"
    />
  </span>
  <span
    class="za-avatar za-avatar--circle za-avatar--string"
    style="color:#2db7f5;background-color:rgb(249, 232, 8)"
  >
    <span
      style="position:absolute;transform:scale(1);left:calc(50% - 4.5px )"
    >
      U
    </span>
  </span>
</div>
`;
